packages/paymasters/src/TokenPaymasterProvider.ts

```diff
@@ -0,0 +1,129 @@
+  token!: PermitInterfaceDAIInstance | PermitInterfaceEIP2612Instance
+  permitSignature!: string
+  protected paymaster!: PermitERC20UniswapV3PaymasterInstance
+  readonly tokenPaymasterInteractor: TokenPaymasterInteractor
+
+  constructor (relayClient: RelayClient, provider: Web3ProviderBaseInterface) {
+    super(relayClient)
+    this.tokenPaymasterInteractor = new TokenPaymasterInteractor(provider)
+  }
+
+  static newProvider (input: TokenPaymasterUnresolvedConstructorInput): TokenPaymasterProvider {
+    if (input.config.maxPaymasterDataLength == null) {
+      input.config.maxPaymasterDataLength = 300
+    }
+    const provider = new TokenPaymasterProvider(new RelayClient(input), input.provider)
+    return provider
+  }
+
+  async init (): Promise<this> {
+    await super.init(true)
+    this.relayClient.dependencies.asyncPaymasterData = this._buildPaymasterData.bind(this)
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this.paymaster = await this.tokenPaymasterInteractor._createPermitERC20UniswapV3Paymaster(this.relayClient.config.tokenPaymasterAddress!)
+    if (this.config.tokenAddress != null) {
+      await this.useToken(this.config.tokenAddress)
+    }
+    return this
+  }
+
+  async _buildPaymasterData (relayRequest: RelayRequest): Promise<PrefixedHexString> {
+    //  Optionally encode permit method,then concatenate token address
+    if (relayRequest.relayData.paymaster !== this.paymaster.address) {
+      throw new Error('Paymaster address mismatch')
+    }
+    const allowance = await this.token.allowance(relayRequest.request.from, relayRequest.relayData.paymaster)
+    let permitMethod = ''
+    if (allowance.eqn(0)) {
+      if (this.config.tokenPaymasterDomainSeparators == null) {
+        throw new Error('TokenPaymasterProvider not initialized. Call init() first')
+      }
+      const domainSeparator: EIP712Domain = this.config.tokenPaymasterDomainSeparators[this.token.address]
+      if (this.permitSignature === PERMIT_SIGNATURE_DAI) {
+        permitMethod = await signAndEncodeDaiPermit(
+          relayRequest.request.from,
+          relayRequest.relayData.paymaster,
+          this.token.address,
+          constants.MAX_UINT256.toString(),
+          this.web3,
+          domainSeparator
+        )
+      } else {
+        permitMethod = await signAndEncodeEIP2612Permit(
+          relayRequest.request.from,
+          relayRequest.relayData.paymaster,
+          this.token.address,
+          constants.MAX_UINT256.toString(),
+          constants.MAX_UINT256.toString(),
+          this.web3,
+          domainSeparator,
+          domainSeparator.version == null ? EIP712DomainTypeWithoutVersion : EIP712DomainType
+        )
+      }
+    }
+    // return abi.encodeParameters(['bytes', 'address'], [permitMethod, this.token.address])
+    return '0x' + removeHexPrefix(this.token.address) + removeHexPrefix(permitMethod)
+  }
+
+  async useToken (tokenAddress: Address): Promise<void> {
```

Comment:
`useXxx` is a name convention used and abused by the React framework for Hooks and given we target web developers with the library we should not call our methods `useSomthing`.
Call it `setTokenAddress` or something.

Reply:
fixed